README.md
# 420 aka Pink Bot
Bot for Telegram that gives a notification at a certain time in all timezones.

## Visuals

![screenshot](Capture.PNG)

## Usage

You add the bot to your channel and then can see all the commands available with the /help command.

## Support

If you have any suggestion, doubt or contribution you can hit me up in Telegram or Twitter @suembra or create an issue in this repository.

## Contributing

If you want to contribute, please check the CONTRIBUTING.md file first.

To run this bot in your localhost, you have to run *npm install* to install all the dependencies and create a file *.env* that has this variables:

ALARM_MINUTE=20
ALARM_HOUR=4
CHAT_ID=your_chat_id
TOKEN=your_bot_token
MAPS_API_KEY=your_maps_api_key

Obviously, you can change the ALARM_MINUTE and ALARM_HOUR variables to the time you want the notifications. If you don't know how to get the chat id you can see some suggestions in https://stackoverflow.com/questions/36099709/how-get-right-telegram-channel-id/45577616 or https://stackoverflow.com/questions/32423837/telegram-bot-how-to-get-a-group-chat-id.

For a Google Maps API key check https://developers.google.com/maps/documentation/maps-static/get-api-key?hl=pt-PT.


## Project status

This project currently has 3 active contributors and you can see the changes and corrections scheduled in the issues tab.


Changelog.md
## [4.2.0] - 23/02/2020

### Added

- When a time notification is ready, the google maps photo is sent and the notification message is attached as the photo's caption, instead of sending 2 separate messages
- 'zoom' parameter in /globe command
- /strain command to give a random strain
- Timezone to time notification
- Wikipedia page of the city in time notification
- MongoDB integration
- /next command gives the city and country

### Fixed

- Added google maps photo to the Islamabad time notification

### Changed

- It is possible to use the /globe command with spaces

## [4.1.5] - 14/02/2020

### Fixed

- Notification on 4:20
- /next command

## [4.1.0] - 13/02/2020

### Fixed

- Google Maps photos

## [4.0.0] - 13/02/2020

### Added

- Command to test bot availability '/ping' to which it replies with 'Pong!' (kaidey)
- Message on the use of the keyword 'depressao' (kaidey)
- '/help' command to give a list of all the commands available
- '/globe' command to give a Google Maps view of a city

### Changed

- Refactoring of all the code to use Typescript (kaidey)
- The 4:20 notification has an image of the city
- The 4:20 notification now has the country of the city

Thank you kaidey for your contributions to this version! You can find him in https://github.com/Kaidey

## [3.0.0] - 12/02/2020

### Added

- Message on mention
- '/next' command to give how much time is left to the next 4:20

### Changed

- Refatoring of the code in multiple files
- Added Riga as a main city in UTC+2
- Message of UTC+2

## [2.0.0] - 11/02/2020

### Added

- '/cities' command that gives the main cities of each timezone

### Fixed

- Fixed the setInterval time to avoid double notifications
- Notification in the alarm hour
- City in UTC+11 timezone was wrong (changed to Norfolk Island)
- Typo in Kiribati (was Kirimati)

## [1.0.0] - 09/02/2020

### Added

- Notifications for the timezones until UTC+16 in the native languages

functions/commands.js
require('dotenv').config()

module.exports = {
    cities: function(bot) {
        bot.onText(/\/cities(.?)/, (msg, match) => {
            const timezone = match.input.replace('/cities', '').trim()
            if (timezone) {
                mainCities(bot, msg.chat.id, timezone)
            } else {
                bot.sendMessage(msg.chat.id, 'What timezone?', {
                    reply_markup: {
                        inline_keyboard: [
                            [{
                                text: 'UTC',
                                callback_data: 'UTC'
                            }],
                            [{
                                text: 'UTC+1',
                                callback_data: 'UTC+1'
                            }],
                            [{
                                text: 'UTC+2',
                                callback_data: 'UTC+2'
                            }],
                            [{
                                text: 'UTC+3',
                                callback_data: 'UTC+3'
                            }],
                            [{
                                text: 'UTC+4',
                                callback_data: 'UTC+4'
                            }],
                            [{
                                text: 'UTC+5',
                                callback_data: 'UTC+5'
                            }],
                            [{
                                text: 'UTC+6',
                                callback_data: 'UTC+6'
                            }],
                            [{
                                text: 'UTC+7',
                                callback_data: 'UTC+7'
                            }],
                            [{
                                text: 'UTC+8',
                                callback_data: 'UTC+8'
                            }],
                            [{
                                text: 'UTC+9',
                                callback_data: 'UTC+9'
                            }],
                            [{
                                text: 'UTC+10',
                                callback_data: 'UTC+10'
                            }],
                            [{
                                text: 'UTC+11',
                                callback_data: 'UTC+11'
                            }],
                            [{
                                text: 'UTC+12',
                                callback_data: 'UTC+12'
                            }],
                            [{
                                text: 'UTC+13',
                                callback_data: 'UTC+13'
                            }],
                            [{
                                text: 'UTC+14',
                                callback_data: 'UTC+14'
                            }],
                            [{
                                text: 'UTC+15',
                                callback_data: 'UTC+15'
                            }],
                            [{
                                text: 'UTC+16',
                                callback_data: 'UTC+16'
                            }]
                        ]
                    }
                })

                bot.on('callback_query', (callbackQuery) => {
                    mainCities(bot, callbackQuery.message.chat.id, callbackQuery.data)
                })
            }
        })
    },

    mentions: function(bot) {
        bot.onText(/(.?)@PinkGlobal420Bot(.?)/, (msg, match) => {
            bot.sendMessage(msg.chat.id, "Fazeçe irmão");
        })
    },

    next: function(bot) {
        bot.onText(/\/next/, (msg, match) => {
            var currentDate = new Date();

            var hourUTC = currentDate.getUTCHours();

            var minutesUTC = currentDate.getUTCMinutes();

            let timeoffset = 0,
                minutesLeft = 0

            if (process.env.ALARM_HOUR < hourUTC) {
                timeoffset = 24 - (hourUTC - process.env.ALARM_HOUR)
            } else if (process.env.ALARM_HOUR > hourUTC) {
                timeoffset = hourUTC - process.env.ALARM_HOUR
            }

            if (minutesUTC < process.env.ALARM_MINUTE) {
                minutesLeft = parseInt(process.env.ALARM_MINUTE) - minutesUTC
            } else {
                minutesLeft = (60 - minutesUTC) + parseInt(process.env.ALARM_MINUTE)
                timeoffset++
            }

            if (timeoffset > 16) {
                minutesLeft = 60 * (24 - timeoffset) - minutesLeft + parseInt(process.env.ALARM_MINUTE)
            }

            nextCity(bot, msg.chat.id, timeoffset, minutesLeft)
        })
    },

    ping: function (bot) {
        bot.onText(/\/ping/, (msg, match) => {
            bot.sendMessage(msg.chat.id, 'Pong!')
        })
    }
}

function mainCities(bot, chatId, timezone) {
    switch (timezone) {
        case 'UTC':
            bot.sendMessage(chatId, 'Main cities: London, Dublin, Lisbon, Abidjan, Accra, Dakar')
            break;
        case 'UTC+1':
            bot.sendMessage(chatId, 'Main cities: Berlin, Rome, Paris, Madrid, Warsaw, Lagos, Kinshasa, Algiers, Casablanca')
            break;
        case 'UTC+2':
            bot.sendMessage(chatId, 'Main cities: RIGA, Cairo, Johannesburg, Khartoum, Kiev, Bucharest, Athens, Jerusalem')
            break;
        case 'UTC+3':
            bot.sendMessage(chatId, 'Main cities: Moscow, Istanbul, Riyadh, Baghdad, Addis Ababa, Doha')
            break;
        case 'UTC+4':
            bot.sendMessage(chatId, 'Main cities: Dubai, Baku, Tbilisi, Yerevan, Samara')
            break;
        case 'UTC+5':
            bot.sendMessage(chatId, 'Main cities: Karachi, Tashkent, Yekaterinburg')
            break;
        case 'UTC+6':
            bot.sendMessage(chatId, 'Main cities: Dhaka, Almaty, Omsk')
            break;
        case 'UTC+7':
            bot.sendMessage(chatId, 'Main cities: Jakarta, Ho Chi Minh City, Bangkok, Krasnoyarsk')
            break;
        case 'UTC+8':
            bot.sendMessage(chatId, 'Main cities: Shanghai, Taipei, Kuala Lumpur, Singapore, Perth, Manila, Makassar, Irkutsk')
            break;
        case 'UTC+9':
            bot.sendMessage(chatId, 'Main cities: Tokyo, Seoul, Pyongyang, Ambon, Yakutsk')
            break;
        case 'UTC+10':
            bot.sendMessage(chatId, 'Main cities: Sydney, Port Moresby, Vladivostok')
            break;
        case 'UTC+11':
            bot.sendMessage(chatId, 'Main cities: Norfolk Island')
            break;
        case 'UTC+12':
            bot.sendMessage(chatId, 'Main cities: Auckland, Suva, Petropavlovsk-Kamchatsky')
            break;
        case 'UTC+13':
            bot.sendMessage(chatId, 'Main cities: Kiribati')
            break;
        case 'UTC+14':
            bot.sendMessage(chatId, 'Main cities: Adak')
            break;
        case 'UTC+15':
            bot.sendMessage(chatId, 'Main cities: Anchorage')
            break;
        case 'UTC+16':
            bot.sendMessage(chatId, 'Main cities: Vancouver')
            break;
        default:
            bot.sendMessage(chatId, 'Invalid timezone or timezone not supported yet')
            break;
    }
}

function nextCity(bot, chatId, timeoffset, minutesLeft) {
    switch (timeoffset) {
        case 0:
            bot.sendMessage(chatId, `As próximas 4:20 são no Porto! Faltam ${minutesLeft} minutos`)
            break;
        case 1:
            bot.sendMessage(chatId, `Les properes 4:20 són a Barcelona! Resten ${minutesLeft} minuts`)
            break;
        case 2:
            bot.sendMessage(chatId, `Nākamie 4:20 nav pat Rīga! ${minutesLeft} minūtes jāiet`)
            break;
        case 3:
            bot.sendMessage(chatId, `4:20 القادمة في دبي! ${minutesLeft} دقيقة للذهاب`)
            break;
        case 4:
            bot.sendMessage(chatId, `اگلا 4:20 اسلام آباد میں ہے! ${minutesLeft} منٹ جانا ہے`)
            break;
        case 5:
            bot.sendMessage(chatId, `Келесі 4:20 - Алматыда! ${minutesLeft} минут жүру керек`)
            break;
        case 6:
            bot.sendMessage(chatId, `4:20 ถัดไปอยู่ในกรุงเทพ! ไป ${minutesLeft} นาที`)
            break;
        case 7:
            bot.sendMessage(chatId, `接下来的4:20在香港！ ${minutesLeft}分钟`)
            break;
        case 8:
            bot.sendMessage(chatId, `次の4:20は東京にあります！あと${minutesLeft}分`)
            break;
        case 9:
            bot.sendMessage(chatId, `The next 4:20 are in Brisbane! ${minutesLeft} minutes to go`)
            break;
        case 10:
            bot.sendMessage(chatId, `The next 4:20 are in Melbourne! ${minutesLeft} minutes to go`)
            break;
        case 11:
            bot.sendMessage(chatId, `The next 4:20 are in Norfolk Island! ${minutesLeft} minutes to go`)
            break;
        case 12:
            bot.sendMessage(chatId, `The next 4:20 are in Auckland! ${minutesLeft} minutes to go`)
            break;
        case 13:
            bot.sendMessage(chatId, `The next 4:20 are in Kiribati! ${minutesLeft} minutes to go`)
            break;
        case 14:
            bot.sendMessage(chatId, `The next 4:20 are in Adak! ${minutesLeft} minutes to go`)
            break;
        case 15:
            bot.sendMessage(chatId, `The next 4:20 are in Anchorage! ${minutesLeft} minutes to go`)
            break;
        case 16:
            bot.sendMessage(chatId, `The next 4:20 are in Vancouver! ${minutesLeft} minutes to go`)
            break;
        default:
            bot.sendMessage(chatId, `The next 4:20 are in Vancouver! ${minutesLeft} minutes to go`)
            break;
    }
}